Add Check Support for Java 21 Record Pattern Syntax: Indentation

child of https://github.com/checkstyle/checkstyle/issues/14890

I have read check documentation:https://checkstyle.org/checks/misc/indentation.html#Indentation
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

---

```
PS D:\CS\test> javac src/Test.java                                              
PS D:\CS\test> cat src/Test.java                                                
record ColoredPoint(boolean p, int x, int c) { }
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }

public class Test {
    void test(Object obj) {
        boolean b = obj instanceof
                ColoredPoint(_,_,_);
        boolean bb = obj instanceof
        ColoredPoint(_,_,_);  // violation
        if (obj instanceof ColoredPoint(_,_,_)) {}
        if (obj instanceof
                ColoredPoint(_,_,_)) {}
        if (obj instanceof
        ColoredPoint(_,_,_)) {  // violation

        }
        if (obj instanceof
ColoredPoint(_,_,_)) {  // violation

        }
        if (obj instanceof
ColoredPoint( // violation
_,  // violation
_,  // violation
_)) // violation
        {

        }

        if (obj instanceof
                Rectangle(
                ColoredPoint(
                        boolean x,
                int y,
                        _),
                ColoredPoint(_,_,_)
                )) {

        }

        if (obj instanceof
Rectangle(  // violation
ColoredPoint(  // violation
        boolean x, // violation
        int y, // violation
        _), // violation
ColoredPoint(_,_,_) // violation
        )) {  // violation

        }
    }

    void testSwitch(Object obj) {
        switch (obj) {
            case Rectangle(
        ColoredPoint _,  // violation
        ColoredPoint _) -> System.out.println("Rectangle");  // violation
            default -> {}
        }
    }
}
PS D:\CS\test> cat config.xml                                                   
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
               <module name="Indentation">

        </module>
    </module>
</module>
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:9:9: 'ColoredPoint' has incorrect indentation level 8, expected level should be 12. [Indentation]
[ERROR] D:\CS\test\src\Test.java:14:9: 'if' child has incorrect indentation level 8, expected level should be 12. [Indentation]
[ERROR] D:\CS\test\src\Test.java:18:1: 'if' child has incorrect indentation level 0, expected level should be 12. [Indentation]
[ERROR] D:\CS\test\src\Test.java:22:1: 'if' child has incorrect indentation level 0, expected level should be 12. [Indentation]
[ERROR] D:\CS\test\src\Test.java:23:1: 'if' child has incorrect indentation level 0, expected level should be 12. [Indentation]
[ERROR] D:\CS\test\src\Test.java:24:1: 'if' child has incorrect indentation level 0, expected level should be 12. [Indentation]
[ERROR] D:\CS\test\src\Test.java:25:1: 'if' child has incorrect indentation level 0, expected level should be 12. [Indentation]
[ERROR] D:\CS\test\src\Test.java:42:1: 'if' child has incorrect indentation level 0, expected level should be 12. [Indentation]
[ERROR] D:\CS\test\src\Test.java:43:1: 'if' child has incorrect indentation level 0, expected level should be 12. [Indentation]
[ERROR] D:\CS\test\src\Test.java:44:9: 'if' child has incorrect indentation level 8, expected level should be 12. [Indentation]
[ERROR] D:\CS\test\src\Test.java:45:9: 'if' child has incorrect indentation level 8, expected level should be 12. [Indentation]
[ERROR] D:\CS\test\src\Test.java:46:9: 'if' child has incorrect indentation level 8, expected level should be 12. [Indentation]
[ERROR] D:\CS\test\src\Test.java:47:1: 'if' child has incorrect indentation level 0, expected level should be 12. [Indentation]
[ERROR] D:\CS\test\src\Test.java:48:9: 'if' child has incorrect indentation level 8, expected level should be 12. [Indentation]
[ERROR] D:\CS\test\src\Test.java:56:9: 'case' child has incorrect indentation level 8, expected level should be 16. [Indentation]
[ERROR] D:\CS\test\src\Test.java:57:9: 'case' child has incorrect indentation level 8, expected level should be 16. [Indentation]
Audit done.
Checkstyle ends with 16 errors.
PS D:\CS\test> 

```

---
Describe what you want in detail.

looks good to me.

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/indentation/IndentationCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/indentation/IndentationCheckTest.java
@@ -3047,6 +3047,50 @@ public class IndentationCheckTest extends AbstractModuleTestSupport {
         verifyWarns(checkConfig, fileName, expected);
     }
 
+    @Test
+    public void testIndentationRecordPattern()
+            throws Exception {
+        final DefaultConfiguration checkConfig = createModuleConfig(IndentationCheck.class);
+        checkConfig.addProperty("forceStrictCondition", "true");
+        checkConfig.addProperty("tabWidth", "4");
+        checkConfig.addProperty("basicOffset", "4");
+        checkConfig.addProperty("braceAdjustment", "0");
+        checkConfig.addProperty("caseIndent", "4");
+        checkConfig.addProperty("throwsIndent", "8");
+
+        final String fileName = getNonCompilablePath(
+                "InputIndentationRecordPattern.java");
+        final String[] expected = {
+            "19:17: " + getCheckMessage(MSG_ERROR, "ColoredPoint", 16, 12),
+            "24:9: " + getCheckMessage(MSG_ERROR, "ColoredPoint", 8, 12),
+            "29:17: " + getCheckMessage(MSG_ERROR, "ColoredPoint", 16, 12),
+            "34:9: " + getCheckMessage(MSG_CHILD_ERROR, "if", 8, 12),
+            "37:1: " + getCheckMessage(MSG_CHILD_ERROR, "if", 0, 12),
+            "39:1: " + getCheckMessage(MSG_CHILD_ERROR, "if", 0, 12),
+            "40:1: " + getCheckMessage(MSG_CHILD_ERROR, "if", 0, 12),
+            "41:1: " + getCheckMessage(MSG_CHILD_ERROR, "if", 0, 12),
+            "42:1: " + getCheckMessage(MSG_CHILD_ERROR, "if", 0, 12),
+            "56:17: " + getCheckMessage(MSG_ERROR, "Rectangle", 16, 12),
+            "57:17: " + getCheckMessage(MSG_ERROR, "ColoredPoint", 16, 12),
+            "58:25: " + getCheckMessage(MSG_ERROR, "boolean", 24, 12),
+            "59:17: " + getCheckMessage(MSG_ERROR, "int", 16, 12),
+            "60:25: " + getCheckMessage(MSG_ERROR, "_", 24, 12),
+            "61:17: " + getCheckMessage(MSG_ERROR, "ColoredPoint", 16, 12),
+            "62:17: " + getCheckMessage(MSG_ERROR, ")", 16, 8),
+            "67:1: " + getCheckMessage(MSG_CHILD_ERROR, "if", 0, 12),
+            "66:1: " + getCheckMessage(MSG_CHILD_ERROR, "if", 0, 12),
+            "68:9: " + getCheckMessage(MSG_CHILD_ERROR, "if", 8, 12),
+            "69:9: " + getCheckMessage(MSG_CHILD_ERROR, "if", 8, 12),
+            "70:9: " + getCheckMessage(MSG_CHILD_ERROR, "if", 8, 12),
+            "71:1: " + getCheckMessage(MSG_CHILD_ERROR, "if", 0, 12),
+            "72:9: " + getCheckMessage(MSG_CHILD_ERROR, "if", 8, 12),
+            "81:13: " + getCheckMessage(MSG_ERROR, ")", 12, 8),
+            "89:9: " + getCheckMessage(MSG_CHILD_ERROR, "case", 8, 16),
+            "90:9: " + getCheckMessage(MSG_CHILD_ERROR, "case", 8, 16),
+        };
+        verifyWarns(checkConfig, fileName, expected);
+    }
+
     private static final class IndentAudit implements AuditListener {
 
         private final IndentComment[] comments;

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationRecordPattern.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationRecordPattern.java
@@ -1,0 +1,100 @@
+//non-compiled with javac: Compilable with Java21                           //indent:0 exp:0
+package com.puppycrawl.tools.checkstyle.checks.indentation.indentation;     //indent:0 exp:0
+
+
+/* Config:                                                                  //indent:0 exp:0
+ * basicOffset = 4                                                          //indent:1 exp:1
+ * braceAdjustment = 0                                                      //indent:1 exp:1
+ * caseIndent = 4                                                           //indent:1 exp:1
+ * tabWidth = 4                                                             //indent:1 exp:1
+ * throwsIndent = 8                                                         //indent:1 exp:1
+ * forceStrictCondition = true                                              //indent:1 exp:1
+ */                                                                         //indent:1 exp:1
+public class InputIndentationRecordPattern {  //indent:0 exp:0
+    record ColoredPoint(boolean p, int x, int c) { } //indent:4 exp:4
+    record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { } //indent:4 exp:4
+
+    void test(Object obj) { //indent:4 exp:4
+        boolean b = obj instanceof //indent:8 exp:8
+                ColoredPoint(_,_,_);  //indent:16 exp:12 warn
+        boolean c = obj instanceof //indent:8 exp:8
+            ColoredPoint(_,_,_);  //indent:12 exp:12
+
+        boolean bb = obj instanceof //indent:8 exp:8
+        ColoredPoint(_,_,_);  //indent:8 exp:12 warn
+
+        if (obj instanceof ColoredPoint(_,_,_)) {} //indent:8 exp:8
+
+        if (obj instanceof //indent:8 exp:8
+                ColoredPoint(_,_,_)) {} //indent:16 exp:12 warn
+        if (obj instanceof     //indent:8 exp:8
+            ColoredPoint(_,_,_)) {} //indent:12 exp:12
+
+        if (obj instanceof  //indent:8 exp:8
+        ColoredPoint(_,_,_)) {  } //indent:8 exp:12 warn
+
+        if (obj instanceof //indent:8 exp:8
+ColoredPoint(_,_,_)) { }  //indent:0 exp:12 warn
+        if (obj instanceof //indent:8 exp:8
+ColoredPoint( //indent:0 exp:12 warn
+_,  //indent:0 exp:12 warn
+_,  //indent:0 exp:12 warn
+_)) {} //indent:0 exp:12 warn
+
+        if (obj instanceof //indent:8 exp:8
+            ColoredPoint( //indent:12 exp:12
+            _,  //indent:12 exp:12
+            _,  //indent:12 exp:12
+            _)) {} //indent:12 exp:12
+
+        if (obj instanceof ColoredPoint( //indent:8 exp:8
+            _,  //indent:12 exp:12
+            _,  //indent:12 exp:12
+            _)) {} //indent:12 exp:12
+
+        if (obj instanceof //indent:8 exp:8
+                Rectangle( //indent:16 exp:12 warn
+                ColoredPoint(  //indent:16 exp:12 warn
+                        boolean x, //indent:24 exp:12 warn
+                int y, //indent:16 exp:12 warn
+                        _), //indent:24 exp:12 warn
+                ColoredPoint(_,_,_)  //indent:16 exp:12 warn
+                ) //indent:16 exp:8 warn
+        ) {} //indent:8 exp:8
+
+        if (obj instanceof //indent:8 exp:8
+Rectangle(   //indent:0 exp:12 warn
+ColoredPoint(   //indent:0 exp:12 warn
+        boolean x,  //indent:8 exp:12 warn
+        int y,  //indent:8 exp:12 warn
+        _),  //indent:8 exp:12 warn
+ColoredPoint(_,_,_)  //indent:0 exp:12 warn
+        )) {}   //indent:8 exp:12 warn
+
+        if (obj instanceof //indent:8 exp:8
+            Rectangle(   //indent:12 exp:12
+            ColoredPoint(   //indent:12 exp:12
+            boolean x,  //indent:12 exp:12
+            int y,  //indent:12 exp:12
+            _),  //indent:12 exp:12
+            ColoredPoint(_,_,_)  //indent:12 exp:12
+            )) {}   //indent:12 exp:8 warn
+
+
+    }  //indent:4 exp:4
+
+    void testSwitch(Object obj) {  //indent:4 exp:4
+        switch (obj) { //indent:8 exp:8
+            case Rectangle( //indent:12 exp:12
+        ColoredPoint _,  //indent:8 exp:16 warn
+        ColoredPoint _) -> System.out.println("Rectangle");  //indent:8 exp:16 warn
+            default -> {} //indent:12 exp:12
+        } //indent:8 exp:8
+        switch (obj) { //indent:8 exp:8
+            case Rectangle( //indent:12 exp:12
+                ColoredPoint _,  //indent:16 exp:16
+                ColoredPoint _) -> System.out.println("Rectangle");  //indent:16 exp:16
+            default -> {} //indent:12 exp:12
+        } //indent:8 exp:8
+    } //indent:4 exp:4
+}//indent:0 exp:0

